design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,290 @@
+# KLIP 12 - Implement High-Availability for Pull queries
+
+**Author**: Vicky Papavasileiou, Vinoth Chandar | 
+**Release Target**: 5.5 | 
+**Status**: _In Discussion_
+**Discussion**: _link to the design discussion PR_
+
+**tl;dr:** Enables high-availability for Ksql pull queries, in case of server failures. Current 
+design for handling failure of an active ksql server, incurs an unavailability period of several 
+seconds (10+) to few minutes due to the Streams rebalancing procedure. This work is based on new 
+Kafka Streams APIs, introduced as a part of [KIP-535](https://cwiki.apache.org/confluence/display/KAFKA/KIP-535%3A+Allow+state+stores+to+serve+stale+reads+during+rebalance), 
+which allow serving stale data from standby replicas to achieve high availability, while providing 
+eventual consistency. 
+
+## Motivation and background
+
+Assume we have a load balancer (LB) and a cluster of three KSQL servers (`A`, `B`, `C`) where `A` 
+is the active server and `B`, `C` are the standbys. `A` receives updates from the source topic 
```

Comment:
The terminology here confuses me. Perhaps I have misunderstood, but, aiui, KSQL servers do not work in an active-passive (aka active-standby) setup. In an active-standby HA setup there is only one server that serves all requests and all other servers are completely passive and just sit there until the active server fails at which point one of them will be promoted to the new active server.
Perhaps I am wrong, but I was under the impression that KSQL servers work in an active-active partitioned setup - i.e. all servers are available to field requests but only one server is the "owner" of a particular key. Can someone clarify?

Reply:
@purplefox ksql uses streaks underneath and it does work in active-passive mode on a partition basis.. i.e a given ksql server will be active for few partitions and just passively replicating state for others.. Overall, all ksql servers would be serving requests for some active partitions at any given time...

Reply:
I wouldn't call that active-passive. Traditionally active-passive means only one server is active at any one time. the other servers literally do nothing at all until they are activated. In an active-passive setup there is no partitioning and everything is replicated to each node.
I'd suggest we use the more standard terminology to avoid confusion :)

Reply:
This is just standard streams terminology. Let's just stick with it :)

Reply:
Do you mean Kafka Streams? I think it would be extremely confusing (especially for outsiders who aren't familiar with the KS redefinitiions) to use such non standard terminology.

Active-Active and active-passive have had well defined meanings in the industry at large for decades. You can google for yourself, e.g.:

https://www.sebastien-han.fr/blog/2012/05/26/differences-between-active-active-and-active-passive-cluster/
https://en.wikipedia.org/wiki/High-availability_cluster

KSQL is active-active with partitioning. Just my 2c

Reply:
@vpapavas what do you think? Might be worth clarifying that we talk about a given partition before we introduce active/standby. and may be a link to Streams HA page for the unfamiliar reader?


`Without loss of generality, we will focus on one partition  for now.`

Reply:
KS certainly uses the terms active and standby on a per-partition basis, so personally I think it's fine to use the same terminology _in the context of a single partition_.

On the flip side, I did need to rely on my existing knowledge of KSQL & KS to understand this paragraph, so it would definitely benefit from improving.  Call out we're talking about a single partition with hot standbys on each of the other nodes.